File: hackerrank/practice/the_coin_change_problem.py
from collections import defaultdict
store = defaultdict(int)
def solve(a, m, n):
	if n == 0:
		return 1
	if n < 0:
		return 0
	if m <= 0 and n >= 1:
		return 0
	x = store[(m-1,n)] if store[(m-1,n)] else solve(a[:-1],m-1,n)
	store[(m-1,n)] = x
	y = store[(m,n-a[m-1])] if store[(m,n-a[m-1])] else solve(a,m,n-a[m-1])
	store[(m,n-a[m-1])] = y
	return x + y
	# return solve(a[:-1],m-1,n)+solve(a,m,n-a[m-1])

n, m  = map(int, raw_input().split())
a = map(int, raw_input().split())
print solve(a,m,n)

File: hackerrank/practice/the_coin_change_problem.cpp
#include <iostream>
#include <map>
#include <utility>

using namespace std;

map<pair<int, int>,long long> store;
long long solve(int a[], int m, int n){
	if (n==0) return 1;
	if (n < 0) return 0;
	if (m <= 0 && n >= 1) return 0;
	long long x = store[make_pair(m-1,n)];
	x = x ? x : solve(a, m-1,n);
	store[make_pair(m-1,n)] = x;
	long long y = store[make_pair(m,n-a[m-1])];
	y = y ? y : solve(a,m,n-a[m-1]);
	store[make_pair(m,n-a[m-1])] = y;
	return x+y;
	// return solve(a, m-1,n)+solve(a,m,n-a[m-1]);
}

int main(void){
	int n, m;
	cin >> n >> m;
	int a[m];
	for (int i = 0; i < m; i += 1){
		cin >> a[i];
	}
	cout << solve(a, m, n) << endl;
}